libjwt/jwt.c

```diff
@@ -1158,3 +1158,215 @@ void jwt_get_alloc(jwt_malloc_t *pmalloc, jwt_realloc_t* prealloc, jwt_free_t *p
+{
+	if (!jwt_valid)
+		return EINVAL;
+
+	*jwt_valid = jwt_malloc(sizeof(jwt_valid_t));
+	if (!*jwt_valid)
+		return ENOMEM;
+
+	memset(*jwt_valid, 0, sizeof(jwt_valid_t));
+	(*jwt_valid)->alg = alg;
+
+	(*jwt_valid)->req_grants = json_object();
+	if (!(*jwt_valid)->req_grants) {
+		jwt_freemem(*jwt_valid);
+		*jwt_valid = NULL;
+		return ENOMEM;
+	}
+
+	return 0;
+}
+
+void jwt_valid_free(jwt_valid_t *jwt_valid)
+{
+	if (!jwt_valid)
+		return;
+
+	json_decref(jwt_valid->req_grants);
+
+	jwt_freemem(jwt_valid);
+}
+
+int jwt_valid_add_required_grant(jwt_valid_t *jwt_valid, const char *grant, const char *val)
+{
+	if (!jwt_valid || !grant || !strlen(grant) || !val)
+		return EINVAL;
+
+	if (get_js_string(jwt_valid->req_grants, grant) != NULL)
+		return EEXIST;
+
+	if (json_object_set_new(jwt_valid->req_grants, grant, json_string(val)))
+		return EINVAL;
+
+	return 0;
+}
+
+int jwt_valid_add_required_grant_int(jwt_valid_t *jwt_valid, const char *grant, long val)
+{
+	if (!jwt_valid || !grant || !strlen(grant))
+		return EINVAL;
+
+	if (get_js_int(jwt_valid->req_grants, grant) != -1)
+		return EEXIST;
+
+	if (json_object_set_new(jwt_valid->req_grants, grant, json_integer((json_int_t)val)))
+		return EINVAL;
+
+	return 0;
+}
+
+int jwt_valid_add_required_grant_bool(jwt_valid_t *jwt_valid, const char *grant, int val)
+{
+	if (!jwt_valid || !grant || !strlen(grant))
+		return EINVAL;
+
+	if (get_js_int(jwt_valid->req_grants, grant) != -1)
+		return EEXIST;
+
+	if (json_object_set_new(jwt_valid->req_grants, grant, json_boolean(val)))
+		return EINVAL;
+
+	return 0;
+}
+
+const char *jwt_valid_get_required_grant(jwt_valid_t *jwt_valid, const char *grant)
+{
+	if (!jwt_valid || !grant || !strlen(grant)) {
+		errno = EINVAL;
+		return NULL;
+	}
+
+	errno = 0;
+
+	return get_js_string(jwt_valid->req_grants, grant);
+}
+
+long jwt_valid_get_required_grant_int(jwt_valid_t *jwt_valid, const char *grant)
+{
+	if (!jwt_valid || !grant || !strlen(grant)) {
+		errno = EINVAL;
+		return 0;
+	}
+
+	errno = 0;
+
+	return get_js_int(jwt_valid->req_grants, grant);
+}
+
+int jwt_valid_get_required_grant_bool(jwt_valid_t *jwt_valid, const char *grant)
+{
+	if (!jwt_valid || !grant || !strlen(grant)) {
+		errno = EINVAL;
+		return 0;
+	}
+
+	errno = 0;
+
+	return get_js_bool(jwt_valid->req_grants, grant);
+}
+
+int jwt_valid_set_now(jwt_valid_t *jwt_valid, const time_t now)
+{
+	if (!jwt_valid)
+		return EINVAL;
+
+	jwt_valid->now = now;
+
+	return 0;
+}
+
+int jwt_valid_set_headers(jwt_valid_t *jwt_valid, int hdr)
+{
+	if (!jwt_valid)
+		return EINVAL;
+
+	jwt_valid->hdr = hdr;
+
+	return 0;
+}
+
+int jwt_valid_del_required_grants(jwt_valid_t *jwt_valid, const char *grant)
+{
+	if (!jwt_valid)
+		return EINVAL;
+
+	if (grant == NULL || !strlen(grant))
+		json_object_clear(jwt_valid->req_grants);
+	else
+		json_object_del(jwt_valid->req_grants, grant);
+
+	return 0;
+}
+
+int jwt_validate(jwt_t *jwt, jwt_valid_t *jwt_valid)
+{
+	int valid = 1;
+
+	// validate algorithm
```

Comment:
Pleaser use this format for here and further comments.

    /* Validate algorithm */